### DISCUSSION TITLE
Node2Vec: Do I actually retrieve embeddings for leaf nodes for a specific graph

### DISCUSSION THREAD

**Discussion by ghost** [pyg-team/pytorch_geometric#4116] [Q&A] [answered]
Hi, 

I am using Node2Vec to embed a graph of which I want to retrieve leaf node embeddings. Now I have a question regarding if my code will actually return the embeddings for the leaf nodes with labels 0, 1 and 2.  For example, I first generate the following graph with the following code:
![Graph](https://user-images.githubusercontent.com/76945330/155322404-8ccf2b0f-eeb2-404e-ab5b-d560e5d362e1.jpg)

```
G = nx.Graph() 
G.add_node(0) 
G.add_node(1)
G.add_node(2) 
G.add_node(3) 
G.add_node(4) 

G.add_edge(3,2)
G.add_edge(3,4)
G.add_edge(4,1)
G.add_edge(4,0)
```
create the edge index with;
```
edge_index = torch.tensor(
                        np.array(G.edges()),
                        dtype=torch.long
                        ).t().contiguous()
```
and use the following to generate embeddings:
```
model_def = Node2Vec(edge_index, embedding_dim=nz, walk_length=20,
             context_size=10, walks_per_node=10,
             num_negative_samples=1, p=1, q=1, sparse=True).to(device)

embeddings = model_def()
```
Then, I want to retrieve the embeddings for the leaf nodes 0, 1 and 2 from the above graph.  I specifically need the leaf node embeddings to have the labels 0, 1 and 2, as I will use these "hierarchical class specific" embeddings later in an ACGAN where the class labels are 0, 1 and 2. 

With the way I generated the graph, will I retrieve the leaf node embeddings with the following code? I.e., does embeddings[1] actually return the embedding for leaf node 1? Or will this return the embedding for the non-leaf nodes 3 and 4, for example? If the latter is the case, how do I need to change the code/graph to be able to retrieve the embeddings for nodes 0, 1 and 2 (without changing the labels of the leaf nodes)? 

```
emb_leafnode0 = embeddings[0]
emb_leafnode1 = embeddings[1]
emb_leafnode2 = embeddings[2]
```
Thank you in advance!

**Comment by rusty1s** [comment, marked as answer]
I'm sorry for the late reply. Yes, the order of embeddings matches with your node indices. As such, `embeddings[0]` will give you the embedding for node zero. Hope that clarifies your concerns :)

**Reply by ghost** [reply]
Thank you very much for your reply!
